proposals/0000-sum-to-tag.md

```diff
@@ -0,0 +1,259 @@
+---
+author: David Feuer
+date-accepted: ""
+ticket-url: ""
+implemented: ""
+---
+
+This proposal is [discussed at this pull request](https://github.com/ghc-proposals/ghc-proposals/pull/369).
+
+# Add `sumToTag#` primop
+
+Add a `sumToTag#` primop indicating which sum alternative is
+represented by a particular unboxed sum value. For example,
+
+```haskell
+sumToTag# (# a | #) = 0#
+sumToTag# (# | | b #) = 2#
+```
+
+## Motivation
+
+It's sometimes useful to get the numerical value of an alternative.
+For example, I might write
+
+```haskell
+newtype Trit# = Trit# (# (##) | (##) | (##) #)
+
+tritToInt# :: Trit# -> Int#
+tritToInt# (Trit# (# (##)|| #)) = 0#
+tritToInt# (Trit# (# |(##)| #)) = 1#
+tritToInt# (Trit# (# ||(##) #)) = 2#
+```
+
+Under the hood, unboxed sums are actually represented by `Int#` tags
+alongside the "payload" values, often with some padding. For example,
+`(# (##) | Bool #)` is represented as something like `(# Int#, Bool #)`.
+The values look like this:
+
+```haskell
+(# (##) | #)   ~=   (# 1#, padding #)
+(# | b #)      ~=   (# 2#, b #)
+```
+
+Note: the choice to make the first alternative `1#` rather than `0#`
+is poorly motivated, and should probably be changed, but that's beyond
+the strict scope of this proposal.
+
+If we could only get our hands on the `Int#` tag, then we could calculate
+`tritToInt#` very efficiently. The C-- code would look much like the
+code that would be produced by
+
+```haskell
+tritToInt'# :: (# Int# #) -> Int#
+tritToInt'# (# t #) = t - 1
+```
+
+Unfortunately, that's *not* the sort of code that `tritToInt#` will
+actually produce. Rather, it will produce code like this:
+
+```haskell
+tritToInt''# :: (# Int# #) -> Int#
+tritToInt''# (# 3# #) -> 2#
+tritToInt''# (# 2# #) -> 1#
+tritToInt''# _ -> 0#
+```
+
+This seems pretty silly, considering that the result of `tritToInt#`
+is likely to be fed straight into some numerical computation. There
+will be extra code and unnecessary conditional branches when all that's
+needed is a simple decrement operation.
+
+## Proposed Change Specification
+
+Add a `sumToTag#` inline primop, analogous to `dataToTag#`. Approximately,
+
+```haskell
+sumToTag# :: forall (xs :: RuntimeRep) (a :: TYPE ('SumRep xs)). a -> Int#
+```
```

Comment:
Why “approximately”? I assume it will be the type of `sumToTag#` in `GHC.Prim`. Just say so.

Reply:
It seems weird to me to call this `forall xs` when it feels a bit more like `foreach xs` or something. But I'm not an expert.

Reply:
The type should be 

```hs
sumToTag# :: forall (xs :: [RuntimeRep]) (a :: TYPE ('SumRep xs)). a -> Int#
```

in order to be well-kinded, right?

Reply:
Yes, I'll fix that.